hal-x86_64/src/mm/mod.rs

```diff
@@ -0,0 +1,831 @@
+    entry: u64,
+    _level: PhantomData<L>,
+}
+
+#[tracing::instrument(level = "info")]
+pub fn init_paging(vm_offset: VAddr) {
+    VM_OFFSET.store(vm_offset.as_usize(), Ordering::Release);
+
+    tracing::info!("initializing paging...");
+
+    let (pml4_page, flags) = crate::control_regs::cr3::read();
+    tracing::debug!(?pml4_page, ?flags);
+    tracing::trace!("old PML4:");
+    let pml4 = PageTable::<level::Pml4>::current(vm_offset);
+
+    // Log out some details about our starting page table.
+    let mut present_entries = 0;
+    for (idx, entry) in (&pml4.entries[..]).iter().enumerate() {
+        if entry.is_present() {
+            tracing::trace!(idx, ?entry);
+            present_entries += 1;
+        }
+    }
+    tracing::trace!(present_entries);
+
+    // Create the recursive pagetable entry.
+    assert!(
+        !pml4.entries[RECURSIVE_INDEX].is_present(),
+        "bootloader must not have used entry 511"
+    );
+    pml4.entries[RECURSIVE_INDEX]
+        .set_present(true)
+        .set_writable(true)
+        .set_phys_addr(pml4_page.base_address());
+    tracing::info!("recursive entry created");
+
+    // Forcibly flush the entire TLB by resetting cr3.
+    unsafe {
+        crate::control_regs::cr3::write(pml4_page, flags);
+    }
+    tracing::info!("new PML4 set");
+
+    // Log out some details about our starting page table.
+    let mut present_entries = 0;
+    for (idx, entry) in (&pml4.entries[..]).iter().enumerate() {
+        if entry.is_present() {
+            tracing::trace!(idx, ?entry);
+            present_entries += 1;
+        }
+    }
+    tracing::trace!(present_entries);
+}
+
+/// This value should only be set once, early in the kernel boot process before
+/// we have access to multiple cores. So, technically, it could be a `static
+/// mut`. But, using an atomic is safe, even though it's not strictly necessary.
+static VM_OFFSET: AtomicUsize = AtomicUsize::new(core::usize::MAX);
+
+impl PageTable<level::Pml4> {
+    fn current(vm_offset: VAddr) -> &'static mut Self {
+        let (phys, _) = crate::control_regs::cr3::read();
+        unsafe { Self::from_pml4_page(vm_offset, phys) }
+    }
+    unsafe fn from_pml4_page(vm_offset: VAddr, page: Page<PAddr, Size4Kb>) -> &'static mut Self {
+        let pml4_paddr = page.base_address();
+        tracing::trace!(?pml4_paddr, ?vm_offset);
+
+        let virt = vm_offset + VAddr::from_usize(pml4_paddr.as_usize());
+        tracing::debug!(current_pml4_vaddr = ?virt);
+        &mut *(virt.as_ptr::<Self>() as *mut _)
+    }
+}
+
+impl<'mapper, A> Map<'mapper, Size4Kb, A> for PageCtrl
+where
+    A: page::Alloc<Size4Kb>,
+{
+    type Handle = Handle<'mapper, level::Pt>;
+    // type Handle =
+    /// Map the virtual memory page represented by `virt` to the physical page
+    /// represented bt `phys`.
+    ///
+    /// # Panics
+    ///
+    /// - If the physical address is invalid.
+    /// - If the page is already mapped.
+    fn map_page(
```

Comment:
These methods should probably be `unsafe`.

Reply:
Hmm, I'm not sure. I think maybe the trait should probably be unsafe to implement, but I think that if the implementation upholds the required invariants, mapping a page is safe?

Reply:
Mapping a page seems unsafe to me. You can do things like aliasing the stack onto the heap, making memory executable, and make read-only memory read-write.

Unmapping a page also seems unsafe to me, because combined with mapping a page, you can break tons of invariants expected by the language.

Reply:
Some of these seem more like an argument for making the APIs for setting _flags_ on an entry unsafe (since that's how you would, e.g. make things writable or whatever)